### PR TITLE
Add a rate limiter that reduce DogStatsD traffic 

### PR DESCRIPTION
**Pull request by ogaca-dd**
### What does this PR do?

Add a rate limiter that reduces the speed at which the Agent reads DogStatsD traffic when the memory usage exceed a threshold.

### Motivation

Avoid the Agent getting OOM killed.

### Describe how to test/QA your changes

On regular host with `hostMemory` total memory.
  1. set `dogstatsd_mem_based_rate_limiter.enabled` to `false`
  2. send high throughput DogStatsD traffic and note the maximum memory usage of the host (`maxMemory`)
  3. set `dogstatsd_mem_based_rate_limiter.enabled` to `true`
  4. compute `memory ratio` = `maxMemory` / `hostMemory`
  5. set `dogstatsd_mem_based_rate_limiter.low_soft_limit` at `0.7` * `memory ratio`
  6. set `dogstatsd_mem_based_rate_limiter.high_soft_limit` at `0.8` * `memory ratio`
  7. send the same DogStatsd traffic as 2. and check memory usage is not greater than `maxMemory`. 

On K8s:
 - configure a cgroup memory limit to a value greater than 700MB
 - send high throughput DogStatsD traffic and check the Agent is not OOM killed

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/dogstatsd/listeners/ratelimit/mem_based_rate_limiter.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	//  SetGCPercent, madvdontneed=1 and debug.FreeOSMemory() can be removed.
+	if goGC > 0 {
+		debug.SetGCPercent(goGC)
+	}
+
+	// madvdontneed=1 forces the memory to be released. This allows the RSS to reflect what
+	// is really used by the Agent
+	if !strings.Contains(os.Getenv("GODEBUG"), "madvdontneed=1") {
+		return nil, errors.New("GODEBUG must have `madvdontneed=1` in order to use this feature")
+	}
```
Comment (djmitche) [inline pkg/dogstatsd/listeners/ratelimit/mem_based_rate_limiter.go:100]: I believe this was only the case through go1.15?

**Review by djmitche — APPROVED** [review]
Does this only apply to UDS?

**Comment by remeh** [comment]
While I understand the why of this pull-request, I'm wondering about the how.

1. This is 100% assuming that all the memory usage of the Agent comes from the UDS+DogStatsD pipeline (i.e. it could stop listening on UDS traffic while the memory usage comes from a Python check)
2. We are capable of actively measuring the RAM usage of the DogStatsD pipelines with all its buffered channels and more or less accurately (based on usages) in the rest of the code (e.g. serialization). I think that the only real "unquantified" RAM usage is the `PacketPool` which theoretically has an unlimited size.
3. I'm also slightly suspicious about the mention of having to manually set `MADVDONTNEED` which is the default in Go since 1.16. What did you test your code with?

In order to limit DogStatsD traffic RAM usage, I think that a better implementation would be to go with a custom `PacketPool` implementation with has a limit in size. WDYT?
Note that I think that this is an interesting approach, which may be helpful as well for the whole Agent.

**Comment by olivielpeau** [comment]
> We are capable of actively measuring the RAM usage of the DogStatsD pipelines with all its buffered channels and more or less accurately (based on usages) in the rest of the code (e.g. serialization). I think that the only real "unquantified" RAM usage is the PacketPool which theoretically has an unlimited size.

About this specifically, the PacketPool isn't bounded per-se, but the design of the dogstatsd pipeline effectively sets a limit on how many packets can be concurrently in-use, because of the channel sizes. The max memory used by `PacketPool`s is pretty much what's described here: https://github.com/DataDog/datadog-agent/blob/main/docs/dogstatsd/internals.md#packetbuffer

**Comment by lebauce** [comment]
@ogaca-dd Thanks for working on this ! We recently introduced a memory controller (https://github.com/DataDog/datadog-agent/blob/main/pkg/util/cgroups/memory_controller.go) that monitors the cgroups limits (through pressure level messages or threshold). It could be interested to integrate your logic into it so that it could be reused in other components

**Comment by ogaca-dd** [comment]
@djmitche 
> Does this only apply to UDS?
Yes. The reason is that in this PR, the Agent doesn't drop payloads, instead it stops reading from the UDS socket. The goal is to apply the back pressure to the client. The client can use a big queue size if they want to not lose metrics or use a small queue size otherwise.

File: pkg/dogstatsd/listeners/ratelimit/geometric_rate_limiter.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		minRate:          config.minRate,
+		maxRate:          config.maxRate,
+		factor:           config.factor,
+		currentRateLimit: config.minRate,
+	}
+}
+
+func (r *geometricRateLimiter) keep() bool {
+	r.tick++
+	if 1/float64(r.tick) <= r.currentRateLimit {
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/dogstatsd/listeners/ratelimit/geometric_rate_limiter.go]: This code is called when receiving an UDS payload which contains hundred of metrics so I don't think it has an impact on the performance. Anyway, I have updated the code as it is very easy to fix.

File: pkg/dogstatsd/listeners/ratelimit/mem_based_rate_limiter.go (hunk 1 above)
Comment (ogaca-dd) [reply pkg/dogstatsd/listeners/ratelimit/mem_based_rate_limiter.go]: Code updated.

**Comment by ogaca-dd** [comment]
@remeh 

> While I understand the why of this pull-request, I'm wondering about the how.
>
> This is 100% assuming that all the memory usage of the Agent comes from the UDS+DogStatsD pipeline (i.e. it could stop listening on UDS traffic while the memory usage comes from a Python check)
>
> We are capable of actively measuring the RAM usage of the DogStatsD pipelines with all its buffered channels and more or less accurately (based on usages) in the rest of the code (e.g. serialization). I think that the only real "unquantified" RAM usage is the PacketPool which theoretically has an unlimited size.

Previous works show there is a strong correlation between the DogStatsD traffic and the memory used by the Agent in practice. This is the reason why this PR is focused on the DogStatsD traffic. The goal is to avoid the Agent to be OOM killed so it is important to take into account all the memories that play a role when an agent is OOM killed (which include the memory used by Python checks).

> I'm also slightly suspicious about the mention of having to manually set MADVDONTNEED which is the default in Go since 1.16. What did you test your code with?

See  https://github.com/DataDog/datadog-agent/pull/10867#discussion_r834182183.

> In order to limit DogStatsD traffic RAM usage, I think that a better implementation would be to go with a custom PacketPool implementation with has a limit in size. WDYT?
Note that I think that this is an interesting approach, which may be helpful as well for the whole Agent.

I don't think it is a better solution as it doesn't count all the memories used by the Agent.

**Comment by ogaca-dd** [comment]
@lebauce 
> @ogaca-dd Thanks for working on this ! We recently introduced a memory controller (https://github.com/DataDog/datadog-agent/blob/main/pkg/util/cgroups/memory_controller.go) that monitors the cgroups limits (through pressure level messages or threshold). It could be interested to integrate your logic into it so that it could be reused in other components

I had a quick look at https://github.com/DataDog/datadog-agent/blob/main/pkg/util/cgroups/memory_controller.go and I think the use cases are different. In my case, I need a pop notification (What is the memory now) whereas `memory_controller.go` provides a push notification.

File: pkg/config/config.go [hunk 3]
```diff
@@ -529,9 +529,10 @@ func InitConfig(config Config) {
 
 	// To enable the following feature, GODEBUG must contain `madvdontneed=1`
 	config.BindEnvAndSetDefault("dogstatsd_mem_based_rate_limiter.enabled", false)
-	config.BindEnvAndSetDefault("dogstatsd_mem_based_rate_limiter.low_soft_limit", 0.6)
+	config.BindEnvAndSetDefault("dogstatsd_mem_based_rate_limiter.low_soft_limit", 0.7)
 	config.BindEnvAndSetDefault("dogstatsd_mem_based_rate_limiter.high_soft_limit", 0.8)
 	config.BindEnvAndSetDefault("dogstatsd_mem_based_rate_limiter.go_gc", 0) // 0 means don't call SetGCPercent
+	config.BindEnvAndSetDefault("dogstatsd_mem_based_rate_limiter.memory_balast", 1024*1024*1024*8)
```
Comment (remeh) [inline pkg/config/config.go:535]: Typo in the configuration field.

File: pkg/dogstatsd/listeners/ratelimit/mem_based_rate_limiter.go [hunk 4]
```diff
@@ -53,6 +58,14 @@ func BuildMemBasedRateLimiter() (*MemBasedRateLimiter, error) {
 		log.Debugf("cgroup limits not detected: %v", err)
 	}
 
+	ballastOnce.Do(func() {
+		ballastSize := config.Datadog.GetInt("dogstatsd_mem_based_rate_limiter.memory_balast")
```
Comment (remeh) [inline pkg/dogstatsd/listeners/ratelimit/mem_based_rate_limiter.go:62]: Nit.

**Review by remeh — CHANGES_REQUESTED** [review]

**Pull request merged**